{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f049f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a39f749b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TTAC', 'GTCT', 'CGGC', 'CCCT', 'GTCA']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from copy import deepcopy\n",
    "import numpy as np\n",
    "\n",
    "from bio.strings import NUC2INT\n",
    "\n",
    "profile = np.array([\n",
    "    [.4, .2, .2, .2],\n",
    "    [0, .6, .2, 0],\n",
    "    [.2, .2, .8, 0],\n",
    "    [0, .2, 0, .8]\n",
    "])\n",
    "\n",
    "Dna = [\n",
    "    \"ttaccttaac\",\n",
    "    \"gatgtctgtc\",\n",
    "    \"acggcgttag\",\n",
    "    \"ccctaacgag\",\n",
    "    \"cgtcagaggt\"\n",
    "    ]\n",
    "Dna = [dna.upper() for dna in Dna]\n",
    "\n",
    "from dataclasses import dataclass\n",
    "from bio.motifs import most_probable_kmer_from_profile\n",
    "\n",
    "@dataclass\n",
    "class Motifs:\n",
    "    kmers: list[str]\n",
    "    profile_with_pseudocount: bool = False\n",
    "\n",
    "    def __post_init__(self) -> None:\n",
    "        self.kmers = [kmer.upper() for kmer in self.kmers] \n",
    "        self.k = len(self.kmers[0])\n",
    "        self.t = len(self.kmers)\n",
    "        self.profile = self.build_profile(self.profile_with_pseudocount)\n",
    "        \n",
    "\n",
    "    def __get_item__(self, idx: int) -> str:\n",
    "        return self.kmers[idx]\n",
    "    \n",
    "    @property\n",
    "    def score(self) -> int:\n",
    "        \"\"\"scores motifs distance to consensus string\"\"\"\n",
    "        distance_to_consensus = 1 - self.profile.max(axis=0)\n",
    "        return int(distance_to_consensus.sum() * self.t)\n",
    "    \n",
    "    # def consensus(self) -> str:\n",
    "\n",
    "    def build_profile(self, with_pseudocounts: bool = False) -> np.ndarray:\n",
    "        \"\"\"creates a frequency-based probability matrix to find each nucleotide (rows) for each position of the k-mer (columns)\"\"\"\n",
    "        profile = np.zeros((4, self.k))\n",
    "        t = self.t\n",
    "        for motif in self.kmers:\n",
    "            for i, c in enumerate(motif):\n",
    "                profile[NUC2INT[c], i] += 1\n",
    "        if with_pseudocounts:\n",
    "            # adding 1 count to each nucleic acid to avoid zero probabilities\n",
    "            profile += 1\n",
    "            t += 4\n",
    "        return profile / t\n",
    "\n",
    "def motifs_from_profile(profile: np.ndarray, Dna: list[str]) -> Motifs: \n",
    "    k = profile.shape[1]\n",
    "    return Motifs([most_probable_kmer_from_profile(dna, k, profile) for dna in Dna])\n",
    "\n",
    "import random\n",
    "\n",
    "def randomized_motif_search(k: int, t: int, Dna: list[str]) -> Motifs:\n",
    "    # random initialization\n",
    "    starting_indices = [random.randint(0, t - k) for _ in range(t)]\n",
    "    motifs = Motifs([Dna[i][j: j+k] for i, j in enumerate(starting_indices)])\n",
    "    best_motifs = deepcopy(motifs)\n",
    "    \n",
    "    while True:\n",
    "        motifs = motifs_from_profile(motifs.profile, Dna)\n",
    "        if motifs.score < best_motifs.score:\n",
    "            best_motifs = deepcopy(motifs)\n",
    "        else:\n",
    "            return best_motifs\n",
    "\n",
    "\n",
    "randomized_motif_search(k=4, t=5, Dna=Dna)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "251409fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.8, 0. , 0. , 0.2],\n",
       "       [0. , 0.6, 0.2, 0. ],\n",
       "       [0.2, 0.2, 0.8, 0. ],\n",
       "       [0. , 0.2, 0. , 0.8]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m2.build_profile()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "77259e2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.4, 0.2, 0.2, 0.2],\n",
       "       [0.2, 0.4, 0.2, 0.2],\n",
       "       [0.2, 0.2, 0.4, 0.2],\n",
       "       [0.2, 0.2, 0.2, 0.4]])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def normalize(ls: list[str]) -> list[str]:\n",
    "    return [s.upper() for s in ls]\n",
    "Dna = [\n",
    "    \"ttaccttaac\", \n",
    "    \"gaTGTctgtc\", \n",
    "    \"ccggcgttag\", \n",
    "    \"cactaacgag\", \n",
    "    \"cgtcagaggt\"]\n",
    "\n",
    "m3 = Motifs([\"taac\", \"gtct\", \"ccgg\", \"acta\", \"aggt\"])\n",
    "m3.profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f4b3b8ca",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "empty range for randrange() (0, -2, -2)",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m args = {\u001b[33m'\u001b[39m\u001b[33mk\u001b[39m\u001b[33m'\u001b[39m: \u001b[32m8\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mt\u001b[39m\u001b[33m'\u001b[39m: \u001b[32m5\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mDna\u001b[39m\u001b[33m'\u001b[39m: \u001b[33m'\u001b[39m\u001b[33mCGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA\u001b[39m\u001b[33m'\u001b[39m}\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[43mrandomized_motif_search\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[22]\u001b[39m\u001b[32m, line 69\u001b[39m, in \u001b[36mrandomized_motif_search\u001b[39m\u001b[34m(k, t, Dna)\u001b[39m\n\u001b[32m     67\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mrandomized_motif_search\u001b[39m(k: \u001b[38;5;28mint\u001b[39m, t: \u001b[38;5;28mint\u001b[39m, Dna: \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mstr\u001b[39m]) -> Motifs:\n\u001b[32m     68\u001b[39m     \u001b[38;5;66;03m# random initialization\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m69\u001b[39m     starting_indices = \u001b[43m[\u001b[49m\u001b[43mrandom\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrandint\u001b[49m\u001b[43m(\u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[43m \u001b[49m\u001b[43m-\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m_\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mrange\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mt\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\n\u001b[32m     70\u001b[39m     motifs = Motifs([Dna[i][j: j+k] \u001b[38;5;28;01mfor\u001b[39;00m i, j \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(starting_indices)])\n\u001b[32m     71\u001b[39m     best_motifs = deepcopy(motifs)\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[22]\u001b[39m\u001b[32m, line 69\u001b[39m, in \u001b[36m<listcomp>\u001b[39m\u001b[34m(.0)\u001b[39m\n\u001b[32m     67\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mrandomized_motif_search\u001b[39m(k: \u001b[38;5;28mint\u001b[39m, t: \u001b[38;5;28mint\u001b[39m, Dna: \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mstr\u001b[39m]) -> Motifs:\n\u001b[32m     68\u001b[39m     \u001b[38;5;66;03m# random initialization\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m69\u001b[39m     starting_indices = [\u001b[43mrandom\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrandint\u001b[49m\u001b[43m(\u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[43m \u001b[49m\u001b[43m-\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m _ \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(t)]\n\u001b[32m     70\u001b[39m     motifs = Motifs([Dna[i][j: j+k] \u001b[38;5;28;01mfor\u001b[39;00m i, j \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(starting_indices)])\n\u001b[32m     71\u001b[39m     best_motifs = deepcopy(motifs)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/lib/python3.11/random.py:362\u001b[39m, in \u001b[36mRandom.randint\u001b[39m\u001b[34m(self, a, b)\u001b[39m\n\u001b[32m    358\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mrandint\u001b[39m(\u001b[38;5;28mself\u001b[39m, a, b):\n\u001b[32m    359\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"Return random integer in range [a, b], including both end points.\u001b[39;00m\n\u001b[32m    360\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m362\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mrandrange\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mb\u001b[49m\u001b[43m+\u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/lib/python3.11/random.py:345\u001b[39m, in \u001b[36mRandom.randrange\u001b[39m\u001b[34m(self, start, stop, step)\u001b[39m\n\u001b[32m    343\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m width > \u001b[32m0\u001b[39m:\n\u001b[32m    344\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m istart + \u001b[38;5;28mself\u001b[39m._randbelow(width)\n\u001b[32m--> \u001b[39m\u001b[32m345\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mempty range for randrange() (\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[33m, \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[33m, \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[33m)\u001b[39m\u001b[33m\"\u001b[39m % (istart, istop, width))\n\u001b[32m    347\u001b[39m \u001b[38;5;66;03m# Non-unit step argument supplied.\u001b[39;00m\n\u001b[32m    348\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m istep > \u001b[32m0\u001b[39m:\n",
      "\u001b[31mValueError\u001b[39m: empty range for randrange() (0, -2, -2)"
     ]
    }
   ],
   "source": [
    "args = {'k': 8, 't': 5, 'Dna': 'CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA'}\n",
    "randomized_motif_search(**args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a1ab7c96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Motifs(kmers=['ACCT', 'ATGT', 'GCGT', 'ACGA', 'AGGT'], profile_with_pseudocount=False)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs_from_profile(m3.profile, normalize(Dna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ba4ddee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 20\n",
      "1 20\n",
      "2 15\n",
      "3 15\n",
      "4 10\n",
      "5 10\n",
      "6 10\n",
      "7 15\n",
      "8 20\n",
      "9 25\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<bound method Motifs.profile of Motifs(kmers=['TTC', 'ATC', 'TTC', 'ATC', 'TTC'])>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bio.motifs import greedy_motif_search, greedy_motif_search_v2, score, build_profile\n",
    "\n",
    "args = {'k': 3, 't': 5, 'Dna': ['GGCGTTCAGGCA', 'AAGAATCAGTCA', 'CAAGGAGTTCGC', 'CACGTCAATCAC', 'CAATAATATTCG']}\n",
    "\n",
    "motifs = greedy_motif_search_v2(**args, with_pseudocounts=True)\n",
    "motifs.profile\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a28679e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 3\n",
      "1 4\n",
      "2 3\n",
      "3 3\n",
      "4 2\n",
      "5 2\n",
      "6 2\n",
      "7 3\n",
      "8 3\n",
      "9 5\n",
      "0 20\n",
      "1 20\n",
      "2 15\n",
      "3 15\n",
      "4 10\n",
      "5 10\n",
      "6 10\n",
      "7 15\n",
      "8 20\n",
      "9 25\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['TTC', 'ATC', 'TTC', 'ATC', 'TTC'], ['TTC', 'ATC', 'TTC', 'ATC', 'TTC'])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1 = greedy_motif_search(**args, with_pseudocounts=True)\n",
    "v2 = greedy_motif_search_v2(**args, with_pseudocounts=True)\n",
    "v1, v2.kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "075529e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([2., 0., 0.])"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bio.motifs import Motifs\n",
    "\n",
    "motifs = Motifs(['TTC', 'ATC', 'TTC', 'ATC', 'TTC'])\n",
    "motifs.score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26f2df42",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAssertionError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[3]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m v2.kmers == v1\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m (build_profile(v1, with_pseudocounts=\u001b[38;5;28;01mTrue\u001b[39;00m) == v2.profile).all()\n",
      "\u001b[31mAssertionError\u001b[39m: "
     ]
    }
   ],
   "source": [
    "assert v2.kmers == v1\n",
    "assert (build_profile(v1, with_pseudocounts=True) == v2.profile).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d20d4eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[2., 0., 0.],\n",
       "        [0., 0., 5.],\n",
       "        [0., 0., 0.],\n",
       "        [3., 5., 0.]]),\n",
       " 10)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2.counts, v2.score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2da303c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([2., 0., 0.])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distance_to_consensus = v2.t - v2.counts.max(axis=0)\n",
    "distance_to_consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "50c794ae",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Motifs' object has no attribute 'consensus'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[17]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mv2\u001b[49m\u001b[43m.\u001b[49m\u001b[43mconsensus\u001b[49m()\n",
      "\u001b[31mAttributeError\u001b[39m: 'Motifs' object has no attribute 'consensus'"
     ]
    }
   ],
   "source": [
    "v2.consensus()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1228bdb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(np.float64(9.0), np.float64(9.0))"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bio.motifs import Motifs\n",
    "\n",
    "motif_exp = Motifs(\"TCTCGGGG CCAAGGTG TACAGGCG TTCAGGTG TCCACGTG\".split())\n",
    "motif_wrong = Motifs(\"AACGGCCA AAGTGCCA TAGTACCG AAGTTTCA ACGTGCAA\".split())\n",
    "\n",
    "motif_exp.score, motif_wrong.score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe173b2a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rosalind-sp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
